File: answers/hist.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <unistd.h>
#include <cmath>

void makeHisto(const cv::Mat &src, cv::Mat &dst, int hist_size, float hist_min, float hist_max) {
    int h = src.size().height;

    // チャンネルをひとまとめにする
    std::vector<cv::Mat> channels;
    cv::split(src, channels);
    cv::Mat tmp;
    for (auto channel : channels) {
        tmp.push_back(channel);
    }

    // 画素数を数える
    cv::Mat hist;
    float range[] = {hist_min, hist_max};
    const float *hist_range = range;
    cv::calcHist(&tmp, 1, 0, cv::Mat(), hist, 1, &hist_size, &hist_range);

    // 正規化
    cv::normalize(hist, hist, 0.0, 1.0, cv::NORM_MINMAX, -1, cv::Mat());

    // 出力画像の生成
    dst = cv::Mat(cv::Size(hist_size+20, h), CV_8UC3, cv::Scalar::all(255));

    // 背景をグレーに描画
    cv::rectangle(dst, cv::Point(10, 10), cv::Point(hist_size+10, h-10), cv::Scalar(240, 240, 240), -1);

    // ヒストグラムを描画
    for (int i=0; i<hist_size; i++) {
        // Fill
        cv::line(
            dst,
            cv::Point(10+i, h-10),
            cv::Point(10+i, h-10-(int)(hist.at<float>(i)*(h-20))),
            cv::Scalar(0, 0, 0),
            1,
            8,
            0
        );

        // 横軸10ずつラインを引く
        if (i%10 == 0) {
            cv::line(dst, cv::Point(10+i, h-10), cv::Point(10+i, 10), cv::Scalar(170, 170, 170), 1, 8, 0);
        }

        // 横軸50ずつ濃いラインを引く
        if (i%50 == 0) {
            cv::line(dst, cv::Point(10+i, h-10), cv::Point(10+i, 10), cv::Scalar(100, 100, 100), 1, 8, 0);
        }
    }
}

File: answers/hist.py
import numpy as np
import cv2
import matplotlib.pyplot as plt

def makeHisto(img, hist_size, hist_min, hist_max):
    h = img.shape[0]

    # チャンネルを一つにまとめる
    img = img.ravel()

    # 画素数を数える
    hist = cv2.calcHist([img], [0], None, [hist_size], [hist_min, hist_max])

    # 正規化
    hist = cv2.normalize(hist, None, 0.0, 1.0, cv2.NORM_MINMAX)

    # 出力画像の生成
    out = np.zeros((h, hist_size+20, 3), np.uint8)
    out += 255

    # 背景をグレーに描画
    out = cv2.rectangle(out, (10, 10), (hist_size+10, h-10), (240, 240, 240), -1)

    # ヒストグラムを描画
    for i in range(hist_size):
        # Fill
        out = cv2.line(out, (10+i, h-10), (10+i, int(h-10-hist[i]*(h-20))), (0, 0, 0), 1, 8, 0)
        # 10ごとにラインを引く
        if i%10 == 0:
            out = cv2.line(out, (10+i, h-10), (10+i, 10), (170, 170, 170), 1, 8, 0)
        # 50ごとにラインを引く
        if i%50 == 0:
            out = cv2.line(out, (10+i, h-10), (10+i, 10), (100, 100, 100), 1, 8, 0)
    return out

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256_dark.png')
    img_orig = cv2.cvtColor(img_orig, cv2.COLOR_BGR2GRAY)
    img_result = makeHisto(img_orig, 256, 0, 256)
    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.waitKey(0)
